### ISSUE TITLE
Grails params missing values with PUT and PATCH after add spring security core

### ISSUE THREAD

**Issue by majaode** [grails-plugins/grails-spring-security-core#554]
### Steps to Reproduce

1. Create app
2. Add spring security core
3. Create controller (User) and action (index) with permitAll
4. Create request with PUT

### Expected Behaviour (Tested with POST method)

Params: [old:Pass, new:Pass, controller:user, format:null, action:index]

### Actual Behaviour (PUT or PATCH)

Params: [controller:user, format:null, action:index]

### Environment Information

- **spring-security-core:** 3.2.2
- **Grails Version:** 3.3.4
- **JDK Version:** 1.8.0_131

### UserController

```package put_test

import grails.plugin.springsecurity.annotation.Secured
import static org.springframework.http.HttpStatus.NO_CONTENT

@Secured(['permitAll'])
class UserController {

    static allowedMethods = [
            index: ["POST", "PUT", "PATCH"]
    ]

    def index() {
    	println "Params: " + params

        render status: NO_CONTENT
    }
}
```

### Sample Request

```
PUT /user/index HTTP/1.1
Host: localhost:8080
Content-Type: application/x-www-form-urlencoded
Cache-Control: no-cache

old=Pass&new=Pass
```

**Comment by ddelponte** [comment]
@majaode , I have been unable to reproduce your issue.  

Everything is working as expected.

1. I verified that `params` work as expected when a `PUT` request is made and the `params` are included in the url
  1. http://localhost:8080/user/index?old=new&new=Pass
1. I also verified that params may be passed as JSON in the body of the `PUT` request and retrieved via a command object

There was also a related discussion at https://github.com/grails/grails-core/issues/9509 that came to the conclusion that similar behavior is not a bug.

Perhaps I'm not properly understanding/reproducing your issue.  If that's the case, would you please attach or link to an example app that duplicates the problem?

Thanks!

**Comment by majaode** [comment]
Hi @ddelponte ,

I use the request body with `application/x-www-form-urlencoded` to send the params. I can't change it to JSON request since the project spec says it should be an `x-www-form-urlencoded` with PUT method. Please check this following request:

`curl -X PUT -H "Content-Type:application/x-www-form-urlencoded" -d 'old=Pass1234&new=Pass1234' http://localhost:8080/user/index`

I expect the same result as POST request. I will upload the sample if you still cannot reproduce the issue.
Please let me know if I missed something.

Thanks.

**Comment by majaode** [comment]
@ddelponte 
So, today I try to run the app with springsecurity is disabled/removed from the dependencies. And the params for PUT and POST are the same. Here are the results:

POST/PUT Request:
```
POST /user/index HTTP/1.1
Host: localhost:8080
Content-Type: application/x-www-form-urlencoded
Cache-Control: no-cache

old=Pass&new=Pass
```

POST Result:
```
Method: POST
Params: [old:Pass, new:Pass, controller:user, format:null, action:index]
```

PUT Result:
```
Method: PUT
Params: [old:Pass, new:Pass, controller:user, format:null, action:index]
```

build.gradle
```
...
dependencies {
    compile "org.springframework.boot:spring-boot-starter-logging"
    compile "org.springframework.boot:spring-boot-autoconfigure"
    compile "org.grails:grails-core"
    compile "org.springframework.boot:spring-boot-starter-actuator"
    compile "org.springframework.boot:spring-boot-starter-tomcat"
    compile "org.grails:grails-web-boot"
    compile "org.grails:grails-logging"
    compile "org.grails:grails-plugin-rest"
    compile "org.grails:grails-plugin-databinding"
    compile "org.grails:grails-plugin-i18n"
    compile "org.grails:grails-plugin-services"
    compile "org.grails:grails-plugin-url-mappings"
    compile "org.grails:grails-plugin-interceptors"
    compile "org.grails.plugins:cache"
    compile "org.grails.plugins:async"
    compile "org.grails.plugins:scaffolding"
    compile "org.grails.plugins:events"
    compile "org.grails.plugins:hibernate5"
    compile "org.hibernate:hibernate-core:5.1.5.Final"
    compile "org.grails.plugins:gsp"
    //compile 'org.grails.plugins:spring-security-core:3.2.2'
    console "org.grails:grails-console"
    profile "org.grails.profiles:web"
    runtime "org.glassfish.web:el-impl:2.1.2-b03"
    runtime "com.h2database:h2"
    runtime "org.apache.tomcat:tomcat-jdbc"
    runtime "com.bertramlabs.plugins:asset-pipeline-grails:2.14.8"
    testCompile "org.grails:grails-gorm-testing-support"
    testCompile "org.grails:grails-web-testing-support"
    testCompile "org.grails.plugins:geb:1.1.2"
    testRuntime "org.seleniumhq.selenium:selenium-chrome-driver:2.47.1"
    testRuntime "org.seleniumhq.selenium:selenium-htmlunit-driver:2.47.1"
    testRuntime "net.sourceforge.htmlunit:htmlunit:2.18"
}
...
```

UserController
```
package put_test

//import grails.plugin.springsecurity.annotation.Secured
import static org.springframework.http.HttpStatus.NO_CONTENT

//@Secured(['permitAll'])
class UserController {

    static allowedMethods = [
            index: ["POST", "PUT", "PATCH"]
    ]

    def index() {
    	println "Method: " + request.method
    	println "Params: " + params

        render status: NO_CONTENT
    }
}
```

Then, try to enable the springsecurity with default setting, and params for PUT request are gone.
I will upload the sample if you still cannot reproduce this issue.

Thanks.

**Comment by ddelponte** [comment]
@majaode , I will continue to investigate this issue, but I do believe I have a work around for you.

The problem appears to be that the `org.springframework.web.filter.HttpPutFormContentFilter.java` filter is not being applied to the request.  That filter fixes the issue where parameters are not being returned for `PUT` requests with a content-type of `application/x-www-form-urlencoded`.

**The Workaround**
1. configure the `HttpPutFormContentFilter` in `grails-app/conf/spring/resources.groovy` by doing the following:

```
import org.springframework.web.filter.HttpPutFormContentFilter

beans = {
    httpPutFormContentFilter(HttpPutFormContentFilter)
}
```

Then register the filter in `Bootstrap.groovy`:

```
SpringSecurityUtils.clientRegisterFilter('httpPutFormContentFilter', SecurityFilterPosition.FIRST)
```

I'm leaning towards registering this filter by default in the spring-security-core plugin so you won't have to manually register it in future releases.

I'll post updates here to keep you informed.

**Comment by majaode** [comment]
@ddelponte , Thanks for your help. Unfortunately, Your solution only work with default JOINED_FILTERS.

### This is works:
```
grails.plugin.springsecurity.filterChain.chainMap = [
		[pattern: '/assets/**',      filters: 'none'],
		[pattern: '/**/js/**',       filters: 'none'],
		[pattern: '/**/css/**',      filters: 'none'],
		[pattern: '/**/images/**',   filters: 'none'],
		[pattern: '/**/favicon.ico', filters: 'none'],
		[pattern: '/v1/**',          filters: 'JOINED_FILTERS']
]
```

### This will produce error:
```
grails.plugin.springsecurity.filterChain.chainMap = [
		[pattern: '/assets/**',      filters: 'none'],
		[pattern: '/**/js/**',       filters: 'none'],
		[pattern: '/**/css/**',      filters: 'none'],
		[pattern: '/**/images/**',   filters: 'none'],
		[pattern: '/**/favicon.ico', filters: 'none'],
		[pattern: '/v1/**',          filters: 'JOINED_FILTERS,-exceptionTranslationFilter']
]
```

### Log
```
2018-07-06 19:43:58.986 ERROR --- [nio-8080-exec-4] g.web.servlet.mvc.GrailsParameterMap     : Error processing form encoded PUT request

java.lang.IllegalStateException: getInputStream() has already been called for this request
	at org.apache.catalina.connector.Request.getReader(Request.java:1233)
	at org.apache.catalina.connector.RequestFacade.getReader(RequestFacade.java:504)
	at grails.web.servlet.mvc.GrailsParameterMap.<init>(GrailsParameterMap.java:90)
	at org.grails.web.servlet.mvc.GrailsWebRequest.getOriginalParams(GrailsWebRequest.java:254)
	at org.grails.web.servlet.mvc.GrailsWebRequest.resetParams(GrailsWebRequest.java:263)
	at org.grails.web.servlet.mvc.GrailsWebRequest.getParams(GrailsWebRequest.java:244)
	at org.grails.web.mime.HttpServletResponseExtension.getMimeTypeForRequest(HttpServletResponseExtension.groovy:127)
	at org.grails.web.mime.HttpServletResponseExtension.getMimeType(HttpServletResponseExtension.groovy:120)
	at org.grails.web.mime.DefaultMimeTypeResolver.resolveResponseMimeType(DefaultMimeTypeResolver.groovy:41)
	at org.grails.web.mapping.mvc.UrlMappingsHandlerMapping.findRequestedVersion(UrlMappingsHandlerMapping.groovy:184)
	at org.grails.web.mapping.mvc.UrlMappingsHandlerMapping.getHandlerInternal(UrlMappingsHandlerMapping.groovy:132)
	at org.springframework.web.servlet.handler.AbstractHandlerMapping.getHandler(AbstractHandlerMapping.java:352)
	at org.springframework.web.servlet.DispatcherServlet.getHandler(DispatcherServlet.java:1160)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:940)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:901)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970)
	at org.springframework.web.servlet.FrameworkServlet.doPut(FrameworkServlet.java:883)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:664)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:742)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:101)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.boot.web.filter.ApplicationContextHeaderFilter.doFilterInternal(ApplicationContextHeaderFilter.java:55)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.boot.actuate.trace.WebRequestTraceFilter.doFilterInternal(WebRequestTraceFilter.java:110)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:127)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:91)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:114)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at javax.servlet.FilterChain$doFilter.call(Unknown Source)
	at grails.plugin.springsecurity.rest.RestTokenValidationFilter.processFilterChain(RestTokenValidationFilter.groovy:121)
	at grails.plugin.springsecurity.rest.RestTokenValidationFilter.doFilter(RestTokenValidationFilter.groovy:87)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at grails.plugin.springsecurity.web.filter.GrailsAnonymousAuthenticationFilter.doFilter(GrailsAnonymousAuthenticationFilter.groovy:54)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:170)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at grails.plugin.springsecurity.rest.RestAuthenticationFilter.doFilter(RestAuthenticationFilter.groovy:139)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at grails.plugin.springsecurity.web.authentication.logout.MutableLogoutFilter.doFilter(MutableLogoutFilter.groovy:64)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at grails.plugin.springsecurity.rest.RestLogoutFilter.doFilter(RestLogoutFilter.groovy:82)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at grails.plugin.springsecurity.web.SecurityRequestHolderFilter.doFilter(SecurityRequestHolderFilter.groovy:58)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.web.filter.HttpPutFormContentFilter.doFilterInternal(HttpPutFormContentFilter.java:103)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:214)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:177)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.grails.web.servlet.mvc.GrailsWebRequestFilter.doFilterInternal(GrailsWebRequestFilter.java:77)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.grails.web.filters.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:67)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:197)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.CorsFilter.doFilterInternal(CorsFilter.java:96)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.boot.actuate.autoconfigure.MetricsFilter.doFilterInternal(MetricsFilter.java:106)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:199)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:504)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:140)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:81)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:87)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:342)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:803)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:790)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1459)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:748)
```

Let me know if I missed anything.
Thanks

**Comment by ddelponte** [comment]
Thanks for the follow up @majaode .  Without an attached example app along with steps to reproduce the issue, I'm having to guess how to fix your problem.

I tested your example locally by adding the filter exception `-exceptionTranslationFilter` in `application.groovy` and it worked as expected.

Rather than registering the `httpPutFormContentFilter` filter first in the chain:

```
SpringSecurityUtils.clientRegisterFilter('httpPutFormContentFilter', SecurityFilterPosition.FIRST)
```

Try adding it a bit later, after the session management filter with:

```
SpringSecurityUtils.clientRegisterFilter('httpPutFormContentFilter', 1750)
```

Please let me know the results.  Thanks!

**Comment by ddelponte** [comment]
@majaode I'm also curious as to why you would exclude the `exceptionTranslationFilter`?

That is the filter that handles any AccessDeniedException and AuthenticationException thrown within the filter chain.

**Comment by ddelponte** [comment]
@majaode , I've also confirmed that the error you reported seeing, `java.lang.IllegalStateException: getInputStream() has already been called for this request`, is not caused by the newly added `HttpPutFormContentFilter`.

That same error is received even when the `HttpPutFormContentFilter` is not added to the filter chain.

**Comment by majaode** [comment]
@ddelponte , I tried to change the order to `1750` and still got error. 
Here is the sample app https://github.com/majaode/grails_put_error

The reason to exclude `-exceptionTranslationFilter` is I am using the Stateless chain in my real project (with Spring Security REST). Back to this issue, these filters below are identical with the one that mentioned in documentation https://grails-plugins.github.io/grails-spring-security-core/latest/
```
grails.plugin.springsecurity.filterChain.chainMap = [
        [pattern: '/assets/**',      filters: 'none'],
        [pattern: '/**/js/**',       filters: 'none'],
        [pattern: '/**/css/**',      filters: 'none'],
        [pattern: '/**/images/**',   filters: 'none'],
        [pattern: '/**/favicon.ico', filters: 'none'],
        [pattern: '/user/**',        filters: 'JOINED_FILTERS,-exceptionTranslationFilter'],
        [pattern: '/**',             filters: 'JOINED_FILTERS,-basicAuthenticationFilter,-basicExceptionTranslationFilter']
]
```

Try with this request:
```
curl -X PUT -H "Content-Type:application/x-www-form-urlencoded" -d 'old=Pass1234&new=Pass1234' http://localhost:8080/user/index
```

**Comment by ddelponte** [comment]
I believe excluding `exceptionTranslationFilter'` from the filter chain is what causes the exception.

```
[pattern: '/user/**',        filters: 'JOINED_FILTERS,-exceptionTranslationFilter'], // <-- -exceptionTranslationFilter' causes error
```

This error is not a result of using the `HttpPutFormContentFilter`.

The workaround would be to not remove `exceptionTranslationFilter' from the filter chain.

**Comment by majaode** [comment]
@ddelponte , thanks for your help.

Remove the `exceptionTranslationFilter ` is actually my goal. I don't want to be redirected to login screen. I have mentioned that I use `Stateless Chain` in my project with (spring security and spring security rest). Please check this out:
Source: http://alvarosanchez.github.io/grails-spring-security-rest/latest/docs/
```
grails.plugin.springsecurity.filterChain.chainMap = [
        //Stateless chain
        [
                pattern: '/**',
                filters: 'JOINED_FILTERS,-anonymousAuthenticationFilter,-exceptionTranslationFilter,-authenticationProcessingFilter,-securityContextPersistenceFilter,-rememberMeAuthenticationFilter'
        ],

        //Traditional, stateful chain
        [
                pattern: '/stateful/**',
                filters: 'JOINED_FILTERS,-restTokenValidationFilter,-restExceptionTranslationFilter'
        ]
]
```
I have also tried to add `authenticationEntryPoint` into `resources.groovy` in order to remove the traditional login redirection.
```
authenticationEntryPoint(org.springframework.security.web.authentication.Http403ForbiddenEntryPoint)
```
In combination with `HttpPutFormContentFilter ` will log the same error as following.
```
...
2018-07-10 11:39:26.551 ERROR --- [nio-8080-exec-2] g.web.servlet.mvc.GrailsParameterMap     : Error processing form encoded PUT request

java.lang.IllegalStateException: getInputStream() has already been called for this request
...
```

**Comment by ddelponte** [comment]
Ok.  Understood.  It looks like we have two separate issues:

1. Grails params missing values with PUT and PATCH after add spring security core
2. Excluding `exceptionTranslationFilter` from filter chain results in error

The first issue above will be resolved by PR https://github.com/grails-plugins/grails-spring-security-core/pull/555 

I'll need to look more into the second issue.

**Comment by ddelponte** [comment]
@majaode I've submitted a PR for the application you referenced above.

https://github.com/majaode/grails_put_error/pull/1

Please test it out and let me know if it satisfies your requirements.

If it looks okay to you, I will update this plugin with the change.

**Comment by ddelponte** [comment]
Fixed by commit be91f75b1a8f39cdd86abc432383083c16469c86
